packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -40,6 +40,59 @@ import { TokenFetchOptions } from "./tokenFetch";
 const afdUrlConnectExpirationMs = 6 * 60 * 60 * 1000; // 6 hours
 const lastAfdConnectionTimeMsKey = "LastAfdConnectionTimeMs";
 
+const localStorageAvailable = isLocalStorageAvailable();
+
+/**
+ * Helper to check the timestamp in localStorage (if available) indicating whether the cache is still valid.
+ */
+function isAfdCacheValid(): boolean {
+    if (localStorageAvailable) {
+        const lastAfdConnection = localStorage.getItem(lastAfdConnectionTimeMsKey);
+        if (lastAfdConnection !== null) {
+            const lastAfdTimeMs = Number(lastAfdConnection);
+            // If we have used the AFD URL within a certain amount of time in the past,
+            // then we should use it again.
+            if (!isNaN(lastAfdTimeMs) && lastAfdTimeMs > 0
+                && Date.now() - lastAfdTimeMs <= afdUrlConnectExpirationMs) {
+                return true;
+            } else {
+                localStorage.removeItem(lastAfdConnectionTimeMsKey);
+            }
+        }
+    }
+
+    return false;
+}
+
+/**
+ * Test if we deal with NetworkErrorBasic object and if it has enough information to make a call
+ * If in doubt, allow retries
+ *
+ * @param error - error object
+ */
+function canRetryOnError(error: any) {
+    // Always retry unless told otherwise.
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+    return error === null || typeof error !== "object" || error.canRetry === undefined || error.canRetry;
+}
+
+/**
+ * Safely tries to write to local storage
+ * Returns false if writing to localStorage fails. True otherwise
+ *
+ * @param key - localStorage key
+ * @returns whether or not the write succeeded
+ */
+function writeLocalStorage(key: string, value: string) {
+    try {
+        localStorage.setItem(key, value);
+        return true;
+    } catch (e) {
+        debug(`Could not write to localStorage due to ${e}`);
```

Comment:
I'd remove debug(). These statements are great for local debugging, but should be removed or converted to telemetry in shipping product. The ideal long term direction - everything is in telemetry, but verboseness is controlled (locally & remotely depending on triggers).

Reply:
Done in latest.